core/src/main/scala/akka/kafka/internal/BatchingFlowStage.scala

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 - 2019 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.kafka.internal
+
+import akka.annotation.InternalApi
+import akka.kafka.CommitterSettings
+import akka.kafka.ConsumerMessage.{Committable, CommittableOffsetBatch}
+import akka.kafka.internal.BatchingFlowStage.FlushableOffsetBatch
+import akka.stream._
+import akka.stream.stage._
+
+import scala.util.{Failure, Success, Try}
+
+/**
+ * INTERNAL API.
+ *
+ * Combined stage for producing, batching commits and committing.
+ */
+@InternalApi
+private[kafka] final class BatchingFlowStage(val committerSettings: CommitterSettings)
+    extends GraphStage[FlowShape[Committable, Try[FlushableOffsetBatch]]] {
+
+  val in: Inlet[Committable] = Inlet[Committable]("FlowIn")
+  val out: Outlet[Try[FlushableOffsetBatch]] = Outlet[Try[FlushableOffsetBatch]]("FlowOut")
+  val shape: FlowShape[Committable, Try[FlushableOffsetBatch]] = FlowShape(in, out)
+
+  override def createLogic(
+      inheritedAttributes: Attributes
+  ): GraphStageLogic = {
+    new BatchingFlowStageLogic(this, inheritedAttributes)
+  }
+}
+
+private final class BatchingFlowStageLogic(
+    stage: BatchingFlowStage,
+    inheritedAttributes: Attributes
+) extends TimerGraphStageLogic(stage.shape)
+    with StageIdLogging {
+
+  import BatchingFlowStage._
+  import CommitTrigger._
+
+  override protected def logSource: Class[_] = classOf[BatchingFlowStageLogic]
+
+  // ---- initialization
+  override def preStart(): Unit = {
+    super.preStart()
+    scheduleCommit()
+    log.debug("BatchingFlowStage initialized")
+  }
+
+  /** Batches offsets until a commit is triggered. */
+  private var offsetBatch: CommittableOffsetBatch = CommittableOffsetBatch.empty
+
+  // ---- Consuming
+  private def consume(offset: Committable): Unit = {
+    log.debug("Consuming offset {}", offset)
+    offsetBatch = offsetBatch.updated(offset)
+    if (offsetBatch.batchSize >= stage.committerSettings.maxBatch) pushDownStream(BatchSize, flush = false)(push)
+    else tryPull(stage.in) // accumulating the batch
+  }
+
+  private def scheduleCommit(): Unit =
+    scheduleOnce(CommitNow, stage.committerSettings.maxInterval)
+
+  override protected def onTimer(timerKey: Any): Unit = timerKey match {
+    case BatchingFlowStage.CommitNow => pushDownStream(Interval, flush = false)(push)
+  }
+
+  private def pushDownStream(triggeredBy: TriggerdBy, flush: Boolean)(
+      emission: (Outlet[Try[FlushableOffsetBatch]], Try[FlushableOffsetBatch]) => Unit
+  ): Unit = {
+    if (activeBatchInProgress) {
+      log.debug("pushDownStream triggered by {} (flush={}, offsetBatch.size={})",
```

Comment:
We should do the commit here instead of propagating the batch downstream. If the user wants to handle upstream stream failures they can handle them on their own.

Reply:
Actually this was my first version of the GraphStage, unfortunately If we do the commit we loose possibility for parallelism.  And implementing parallelism manually in a `GraphStage` seems like a lot more complex than combining this stage with  simple follow-up steps in the `Committer`. 

Since `Committer` is the public API and not this Graph I assumed this as acceptable solution.